common/persistence/cassandra/metadata_store.go

```diff
@@ -198,6 +202,110 @@ func (m *MetadataStore) UpdateNamespace(request *p.InternalUpdateNamespaceReques
 	return nil
 }
 
+func (m *MetadataStore) RenameNamespace(request *p.InternalRenameNamespaceRequest) error {
+	return m.RenameNamespace1(request)
+}
+
+// Does almost everything in one batch. The only separate thing is namespaecs_by_id update which is almost not used now.
+// Can be retried but data between steps is inconsistens (points to the name which doesn't exist).
+func (m *MetadataStore) RenameNamespace1(request *p.InternalRenameNamespaceRequest) error {
+	// Step 1. Update `namespaces_by_id` row to new name.
+	// Step 2. Batch of:
+	//         Insert row into `namespaces` table with new name and new `notification_version`.
+	//         Delete row from `namespaces` table with old name.
+	//         Update `notification_version` in metadata row
+
+	if updateErr := m.session.Query(templateUpdateNamespaceByIdQuery,
+		request.Name,
+		request.Id,
+	).Exec(); updateErr != nil {
+		return serviceerror.NewUnavailable(fmt.Sprintf("RenameNamespace operation failed to update 'namespaces_by_id' table. Error: %v", updateErr))
+	}
+
+	batch := m.session.NewBatch(gocql.LoggedBatch)
+	batch.Query(templateCreateNamespaceByNameQueryWithinBatchV2,
+		constNamespacePartition,
+		request.Id,
+		request.Name,
+		request.Namespace.Data,
+		request.Namespace.EncodingType.String(),
+		request.NotificationVersion,
+		request.IsGlobal,
+	)
+	batch.Query(templateDeleteNamespaceByNameQueryV2,
+		constNamespacePartition,
+		request.PreviousName,
+	)
+	m.updateMetadataBatch(batch, request.NotificationVersion)
+
+	previous := make(map[string]interface{})
+	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	if err != nil {
+		// TODO: Update namepspaces_by_id to previous name before returning error???
+		return serviceerror.NewUnavailable(fmt.Sprintf("RenameNamespace operation failed. Error: %v", err))
+	}
+	defer func() { _ = iter.Close() }()
+
+	if !applied {
+		// TODO: Update namepspaces_by_id to previous name before returning error???
+		return serviceerror.NewUnavailable(fmt.Sprintf("RenameNamespace operation failed because of conditional failure."))
```

Comment:
this should be an CAS related error?

Reply:
yes, `notification_version` in metadata row is also update there.